### DISCUSSION TITLE
Индикаторы для Quik

### DISCUSSION THREAD

**Discussion by Mingary** [nick-nh/qlua#8] [Ideas]
Еще раз здравствуйте, прошу прощения, что опубликовал свой вопрос не в том разделе, не сразу разобрался.
Да, индикаторы имеют открытый код, насколько возможно реализовать в Quik - мне трудно сказать, я не слишком разбираюсь в тонкостях программирования... 
Прикрепляю один из индикаторов, чтобы Вы могли оценить, можно ли что-то сделать.

Спасибо, что откликнулись 
[TMMS.ZIP](https://github.com/nick-nh/qlua/files/8596469/TMMS.ZIP)
!

**Comment by nick-nh** [comment]
Да, его можно сделать. Просто еще один осциллятор.

**Reply by Mingary** [reply]
Вы возьметесь ? Сколько времени это может занять ?

**Reply by nick-nh** [reply]
Когда будет время. Если он Вам нужен срочно, то могу, конечно, сделать на заказ.
Также необходимо определить нужен ли только сам алгоритм гистограммы или еще доп. индикаторы, что я вижу в нем, например Stochastic Momentum Index. Он, судя по всему, формирует доп. линии поверх основной гистограммы.
Также в нем есть алерты, что, конечно, можно сделать, но если они очень нужны.

**Reply by Mingary** [reply]
Хорошо, я буду ждать, когда сможете сделать, буду Вам очень признателен. 

Те дополнительные индикаторы, которые есть в этом - да, они необходимы. А вот алертами можно и пожертвовать, наверное, реализовать их на Lua не самая простая задача, тогда и не стоит тратить на них силы. Если удастся сохранить стрелки, которые рисует на диаграмме Stochastic_Momentum_Index, то это очень хорошо, если нет - пусть будет без них.

С Вашего позволения, я прикреплю еще несколько остальных индикаторов, которые, на мой непрофессиональный взгляд, должны оказаться чуть попроще. Пожалуйста, не сочтите за наглость, просто они все равноценны и одинаково необходимы..
Относительно них точно также - алерты не столь важны и необходимы, а вот все остальное касательно отрисовки - да, очень нужно.
Быть может, эти индикаторы и Вам сослужат службу или пригодятся.

Большое спасибо за помощь !
[Other indies.zip](https://github.com/nick-nh/qlua/files/8596626/Other.indies.zip)

**Reply by nick-nh** [reply]
Если удастся сохранить стрелки, которые рисует на диаграмме Stochastic_Momentum_Index

Если он рисуется в виде отдельной диаграммы в отдельном окне, то для Квика это будет отдельный индикатор. Квик не умеет в одном индикаторе рисовать в нескольких окнах. Т.к. Вы не привели пример окна, то я и не вижу как он рисуется в MT4.

**Reply by Mingary** [reply]
Отрисовываются индикаторы TMMS Oscillator и Stochastic Color Classic вот так.
Получается тогда, что TMMS Oscillator в Квике разъедется на 2 отдельных окна (подвала) из-за стрелок ?

![Indi_pic](https://user-images.githubusercontent.com/104645616/166101470-6c8e910b-e77f-45a8-b1ad-24f084a3d987.jpg)

**Reply by nick-nh** [reply]
Если рисуются в одном окне, то нет, можно вывести все в одном индикаторе.

**Reply by Mingary** [reply]
В терминале MT4 да, все элементы индикатора рисуются в одном и том же окне.

**Reply by nick-nh** [reply]
Но вот нарисовать линии с изменением цвета, как в Stochastic Color - уже Квик не может. Только если рисовать линию через точки, тогда можно будет выводить точки разных цветов. Так пришлось делать для индикатора THV coral.
Также Квик не может нарисовать свои бары, поэтому Heikin Ashi тоже не нарисовать.

**Reply by Mingary** [reply]
Если можно задавать цвет точек в зависимости от их местоположения в зоне индикатора (то есть, в зоне 80+ - зеленые, в зоне 20- - красные), то пусть так. Наверное, плотность точек может быть высокой, чтобы практически сливаться в линию, создавая полноценный эффект ? Я, к сожалению, не смог запустить в своем Квике Ваш Коралл почему-то, ну да неважно.
А насчет Synergy APB - можно написать, чтоб хоть из подвала, но работал ? По подобию Heikin Ashi.

**Reply by nick-nh** [reply]
В чем проблема запуска THV?

**Reply by Mingary** [reply]
А я даже не знаю, все Ваши индикаторы нормально отображаются в перечне LuaIndicators, а он просто не появляется даже в списке доступных. Может. что-то я криво делал, но дважды попробовал, и не появился.

**Reply by nick-nh** [reply]
Проверьте кодировку файла.

**Reply by Mingary** [reply]
Да, спасибо, попробую. Наверное, дело в ней, потому что код THV на GitHub`е отображается иначе, чем код остальных индикаторов (визуально - вся страница кода раскрашена цветными полями :) )

**Reply by nick-nh** [reply]
Может сделать файл в кодировке 1251 и просто через буфер обмена вставить текст.

**Reply by Mingary** [reply]
Я копировал код со страницы, вставлял в блокнот и сохранял с расширением "lua". И удалось так сделать со всеми индикаторами, которые я брал, кроме THV :)

**Reply by Mingary** [reply]
Здравствуйте ! Можно поинтересоваться, получается ли что-нибудь с индикаторами, которые я сбрасывал ?

**Reply by nick-nh** [reply]
Добрый день. В Порядке очереди. Дойдут руки, сделаю. Там ничего особенного.

**Reply by nick-nh** [reply]
Добрый день. Добавил индикатор https://github.com/nick-nh/qlua/blob/master/smTMMS.lua
Только обращаю внимание, что он использует maLib.lua. Ее необходимо расположить рядом с индикатором.

**Reply by Mingary** [reply]
Здравствуйте !
Огромное Вам спасибо! Вы большой молодец !
Только мне, к сожалению, на удалось запустить индикатор в своем Квике...
Я пробовал сохранять код в разных кодировках, поместил maLib.lua в папку с индикаторами но smTMMS.lua так и не появился в окне индикаторов, доступных к добавлению на график... Я в растерянности, больше не знаю, что попробовать сделать..
Может быть, если Вас не слишком затруднит, у Вас найдется возможность отправить мне на почту alaei@yandex.ru файлы "maLib.lua" и " smTMMS.lua" в виде готовых lua-индикаторов (чтобы исключить вероятность того, что я при сохранении кода делаю что-то не так) ?

**Reply by nick-nh** [reply]
Исправил кодировку.

**Reply by Mingary** [reply]
Прошу прощения, к сожалению, мой Квик по-прежнему его не видит...Теряюсь в догадках, в чем может быть дело. Код индикатора сохранен в 1251 ANSI, делал несколько раз, перезагружал Квик, перезагружал машину, очищал папку Luaindicators  от всех индикаторов и оставлял только TMMS. Ничего из этого не сработало, индикатор так и не появился в списке доступных..

**Reply by Mingary** [reply]
Ситуация, кстати, такая же, как и с Кораллом ) Тот тоже не появился. что я ни делал

**Reply by nick-nh** [reply]
Какая версия Квика? В 9.0-9.2 версиях ошибка, не позволяющая использовать индикаторы с сервисными функциями глобального контекста.

**Reply by nick-nh** [reply]
Если это так, то можете убрать определение цвета линий

local isDark        = _G.isDarkTheme()
local line_color    = isDark and RGB(240, 240, 240) or RGB(0, 0, 0)

Например, просто определив какой-то цвет, убрав вызов isDarkTheme()

local line_color    = RGB(0, 0, 0)

**Comment by Mingary** [comment]
Версия Квика - 8.7. Ошибки, по сути, никакой нет, то есть, Квик ничего не выдаёт. Просто индикатор не появляется в перечне выбора и все. Я копирую его в папку Luaindicators с расширением “lua”, но он просто не появляется в перечне доступных индикаторов в Квике, как будто файла не существует, или я не помещал его в папку Luaindicators.

**Reply by nick-nh** [reply]
Это происходит потому что в момент добавления Квик компилирует все скрипты из папки. Те что не смог, не показывает.

Можете этот индикатор добавить в доступные скрипты и запустить. Он покажет ошибку компиляции.

**Reply by Mingary** [reply]
Попробовал запустить индикатор как скрипт.
Вот какую ошибку выдал Quik:

QUIK_x64\scripts\TMMS.lua:3: attempt to call a nil value

**Reply by nick-nh** [reply]
Вы изменили текст индикатора? В третей строке комментарий. Если убрали его, то третья значимая строка - это подключение maLib.

**Reply by Mingary** [reply]
Все заработало ! Прошу прощения, что вводил Вас в заблуждение, дело оказалось в библиотеке maLib. 
Я решил еще раз заменить ее, взял снова код с сайта и обновил. После этого (ура!) индикатор появился в перечне и стал отображаться. Насколько пока могу судить, работает исправно и корректно ! Огромное Вам спасибо еще раз, отличная работа ! Лично для меня, далекого от программирования - вообще за гранью возможности )

**Reply by Mingary** [reply]
Вы не будете на меня обижаться, если я попрошу Вас сделать для Квика два вот этих индикатора, если это возможно?
Не хочу выглядеть навязчивым, еще хуже наглым, Ваш талант - единственный для меня способ сделать Квик юзабельным относительно опыта MT4 )

[triline.zip](https://github.com/nick-nh/qlua/files/8778395/triline.zip)
![Triline](https://user-images.githubusercontent.com/104645616/170473221-cd6d932e-6f8a-4b0b-ac00-8b3043d02f61.jpg)

**Reply by nick-nh** [reply]
Чуть позже. Хотя Timing Indicator уж очень сильно напоминает THV.

**Reply by Mingary** [reply]
Спасибо Вам большое за помощь ! Я буду ждать )

**Comment by nick-nh** [comment]
Добрый день.

Добавил два индикатора
JMA_Star_Light
JRSX

**Comment by madby31** [comment]
Добрый день!
У Вас кажется ошибка в индикаторе BidAskVol.lua в строках 346 и 380. Вместо Delta[index] наверное должно быть CumDelta[index] ?
А вообще, большое спасибо за индикатор, очень упростил мне задачу.

**Reply by nick-nh** [reply]
Да, действительно. Исправил.

**Reply by madby31** [reply]
Сегодня тестирую в терминале. Первые замечания. Кумулятивная дельта "цепляет" значение дельты с предыдущей сессии в 23:49.
Второе. Некорректно отображается текущие значения дельты. Первый снимок - текущая картина, второй снимок - после удаления и повторного выставления индикатора. После перезапуска, историю просчитывает правильно, а потом беда. Значения последних баров (не текущего) пересчитываются. Значения проверяю по другому индикатору, расчеты которого проверял вручную.
![Скриншот 25-08-2022 10 16 59](https://user-images.githubusercontent.com/41211955/186600300-d441a4a8-9fa1-4f24-bea1-cee2099d47f6.png)
![Скриншот 25-08-2022 10 17 49](https://user-images.githubusercontent.com/41211955/186600330-537d1fca-b320-4df0-af28-232fccf1d8e7.png)

**Reply by nick-nh** [reply]
А почему не брать данные с вечерней сессии. Это просили сделать специально - вот и сделано. Впрочем, можете отключить - это не сложно.

По поводу второго - позже проверю.
Возможно опять функция SearchItems сбоит. Был один вариант без нее - простой перебор сделок. Хотя, добавил на график Si, сравнение объема бара и суммы двух величин пока дают корректное значение для онлайн баров.

**Reply by madby31** [reply]
Не дописал важное, на снимках - только кумулятивная дельта в виде гистограммы. ТОС не хранит данных с предыдущей сессии, поэтому один последний бар последний вносит погрешность на текущую сессию.

**Reply by nick-nh** [reply]
Как это не хранит. Хранит.

![изображение](https://user-images.githubusercontent.com/11758539/186610045-2685622e-5d41-4e0b-a404-ce6ff1eba493.png)

**Reply by madby31** [reply]
Извините, ТОС тупо в начало промотал, а там данные с 6 утра. Спасибо, буду разбираться, по крайней мере база есть.

**Reply by madby31** [reply]
Кажется нашел, почему кумулятивная дельта перерисовывается:
В строке 334:
`CumDelta[ind] 	= CumDelta[ind] + localDelta*delta_koeff`
наверное должно быть:
`CumDelta[ind] 	= CumDelta[ind-1] + localDelta*delta_koeff`
По крайней мере, у меня так работает нормально.

**Reply by nick-nh** [reply]
Обратите внимание на строку 301
CumDelta[index] 		= CumDelta[index-1] or 0

При каждом обращении текущее значение устанавливается по прошлому бару. А в функции calc_delta к нему добавляется дельта текущего бара. Так что это равносильно.

**Comment by madby31** [comment]
Добрый день!
В индикаторе Extremums файл лога перезаписывается при пересчете индикатора.
Предлагаю следующее улучшение кода:
Строку 7 удалить.
После строки 19 добавить код:
`local LogFileName = _G.getWorkingFolder().."\\LuaIndicators\\Extremums\\"..tostring(os.date("%Y-%m-%d",os.time())).."_"...."_Extremums.txt" -- файл лога создается каждый день`
Функцию MyLog в следующем виде:
```
local function myLog(...)
    if logFile==nil then return end
    logFile = io.open(LogFileName,"r+");
    if (logFile==nil) then
        logFile = io.open(LogFileName,"w"); 
    end	
    logFile:seek("end",0)
    logFile:write(tostring(os.date("%c",os_time())).." "..log_tostring(...).."\n");
    logFile:flush();
    logFile:close()
end
```

**Comment by nick-nh** [comment]
Лог нужен только для отладки. В обычном режиме он не пишется, поэтому задачи его сохранять не было вовсе. Каждый раз открывать файл - не лучшее решение.

**Reply by madby31** [reply]
А можно ссылку на описание этого индикатора, наверняка брали откуда-то исходные данные. Логирую сигналы для анализа, но пока не понимаю, почему на фьючерсе РСТ обновляются редко, а на РТС мини часто.

**Reply by nick-nh** [reply]
Этот индикатор писался по описанию клиента.
Смысл очень простой: если 3 бара на интервале достигают одной той же точки по вершинам и этот уровень не пробитый внутри интервала, то сигнал.

**Reply by madby31** [reply]
Спасибо!
